plat/common/arm/mte.c

```diff
@@ -0,0 +1,117 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) 2022, Michalis Pappas <mpappas@fastmail.fm>.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <arm/arm64/cpu.h>
+#include <stddef.h>
+#include <uk/plat/mte.h>
+#include <uk/arch/lcpu.h>
+#include <uk/arch/random.h>
+#include <uk/arch/traps.h>
+#include <uk/event.h>
+
+void *mte_tag_contig(void *ptr, size_t size)
+{
+	uint64_t addr = (uint64_t)ptr;
+
+	UK_ASSERT(!(size % MTE_TAG_GRANULE));
+
+	addr = mte_insert_random_tag(addr);
+
+	for (unsigned int i = 0; i < size; i += MTE_TAG_GRANULE)
+		mte_store_alloc(addr + i);
```

Comment:
Using an `unsigned int` is probably not a good idea here. Also, the definition should be outside the loop, and you can use `__vaddr_t` as type for a virtual address.

```suggestion
	__vaddr_t addr = (__vaddr_t)ptr;
	__off off;
	
	UK_ASSERT(!(size % MTE_TAG_GRANULE));

	addr = mte_insert_random_tag(addr);

	for (off = 0; off < size; off += MTE_TAG_GRANULE)
		mte_store_alloc(addr + off);
```
Also add the `#include <uk/arch/types.h>` for `__vaddr_t` and `__off`

Reply:
> Using an unsigned int is probably not a good idea here. 

Okay, but then why `off_t` and not `size_t`, which also the type of `size` that the loop compares against?

> Also, the definition should be outside the loop

Actually I think it's a pretty good practice to do that, as it reduces clutter in the variable declarations at the top of the function, so I would like if we considered to permit (or even mandate) this practice in the guidelines. Do you see any downsides?

> and you can use __vaddr_t as type for a virtual address.

The rationale for using `uint64_t` is that this feature is not expected to be portable as it is only available in AArch64. I make this assumption in general for certain features in `arm64`, while for the parts that I want to be backwards compatible,  I tend to use generic types (eg in the implementation of the SMCCC that also applies to Armv7, and both AArch64 and AArch32 in Armv8). Unless we want to be 128-bit ready, that is 😄

Reply:
> Okay, but then why off_t and not size_t, which also the type of size that the loop compares against?

Right, `off_t` would not make sense. But `__off` is defined in `uk/arch/types.h` along with `__u64`, `__sz` etc. and is one of those Unikraft native types.

> Actually I think it's a pretty good practice to do that, as it reduces clutter in the variable declarations at the top of the function, so I would like if we considered to permit (or even mandate) this practice in the guidelines. Do you see any downsides?

I see your point and I also had this feeling a few times that the loop variable felt like clutter in the definition block. However, I am a bit reluctant to add the exception for loop variables to the coding convention as it makes the convention more complex and opens the door for further exceptions. For example, you could argue why not having a general exception for variables used only in a nested/closed scope (which is a valid point, btw.). Making it mandatory is too late, from my perspective, as basically no code in UK would follow this. Nevertheless, we can discuss this topic as part of the coding conventions. For now, I think, since all other code has the definition outside, I would go with that.

> The rationale for using uint64_t is that this feature is not expected to be portable as it is only available in AArch64. I make this assumption in general for certain features in arm64...

Ok, makes sense. Then maybe use `__64` if you want to keep it consistent with `__sz`.

Reply:
BTW: I added a section about variable definitions and a comment with your feedback related to loop variables to the code conventions doc so we can discuss.

Reply:
ACK, that shold be `__u64` indeed, will change.

> Right, off_t would not make sense. But __off is defined in uk/arch/types.h along with __u64, __sz etc. and is one of those Unikraft native types.

Sorry I think I made things confusing. I meant `__off` is the native equivalent of `off_t`, but that type doesn't make sense conceptually in this case, as AFAIK it is normally used for files. So the `__sz` type currently used seems fine to me for this case.

Reply:
ok, that is fine, too 😃 👍🏻